torch_xla/csrc/tensor.cpp

```diff
@@ -1430,31 +1448,58 @@ std::shared_ptr<XLATensor::Async> XLATensor::ScheduleSyncTensorsGraph(
     ComputationCache::TypePtr cached_computation) {
   tensorflow::profiler::TraceMe activity(
       "ScheduleSyncTensorsGraph", tensorflow::profiler::TraceMeLevel::kInfo);
+  bool is_sharded = cached_computation->is_sharded;
+  std::cout << "ScheduleSyncTensorsGraph, is_sharded: " << is_sharded
+            << std::endl;
+  auto computation = Computation(cached_computation->computation);
+  std::cout << "- " << computation.to_string() << std::endl;
   TensorCollectionBarrier(coll);
   std::shared_ptr<Async> async = std::make_shared<Async>(
       coll, std::move(parameters_data), std::move(tensors_data),
       std::move(cached_computation));
 
-  auto syncfn = [async, hash = coll->hash]() {
+  auto syncfn = [async, is_sharded, hash = coll->hash]() {
     xla::ComputationClient::ExecuteComputationOptions options;
     try {
-      TF_VLOG(3) << "Executing IR graph hash "
-                 << torch::lazy::HashToString(hash) << " on device "
-                 << async->device << " ...";
-      auto results = xla::ComputationClient::Get()->ExecuteComputation(
-          *async->cached_computation->computation,
-          UnwrapXlaData(async->parameters_data), async->device, options);
-      TF_VLOG(3) << "Executing IR graph hash "
-                 << torch::lazy::HashToString(hash) << " on device "
-                 << async->device << " done!";
-
+      std::vector<xla::ComputationClient::DataPtr> results;
+      // TODO(yeounoh) Execute replicated if the compiled computation
+      // is partitioned.
+      if (is_sharded) {
+        std::vector<std::string> devices =
+            xla::ComputationClient::Get()->GetAllDevices();
+        std::vector<std::vector<xla::ComputationClient::DataPtr>>
+            device_arguments = torch_xla::ShardingUtil::InputHandler(
+                UnwrapXlaData(async->parameters_data), devices);
+        xla::ComputationClient::ExecuteReplicatedOptions execute_options;
+        execute_options.explode_tuple = options.explode_tuple;
+
+        TF_VLOG(3) << "Executing IR graph hash "
+                   << torch::lazy::HashToString(hash) << " on all devices.";
+        results = xla::ComputationClient::Get()->ExecuteReplicated(
```

Comment:
hmm, `results` here is `virtual_data_ptr`? I saw done below we do `UnwrapXlaData(async->tensors_data[i])->Assign(*results[i])` so each element in `result` here should represent a tensor in a virtual device, hence should be a `virtual_data_ptr`, is my understanding correct?

Reply:
Executed results are not virtual and reside on actual devices. The scope of virtual device is to defer physical device data creation until we are certain about the sharding strategies, so we can avoid sending the full unpartitioned data before sending partitioned data again. Do we want to expand the notion of "virtual" device and its usage in our codebase?

Reply:
I am confuse, `ExecuteReplicated` returns a ` virtual std::vector<std::vector<DataPtr>>` according to 
https://github.com/pytorch/xla/blob/3db9008f92a203c202299355d13b9f2c8bba2175/third_party/xla_client/computation_client.h#L263-L267

but you are assign it to a `std::vector<xla::ComputationClient::DataPtr>`. How does that works?

Reply:
It takes the first (arbitrary, since it's replicated) element of the nested vector output.

Reply:
oh ok..

Reply:
can you add a log? It is OK to do that in a separate pr.